### PR TITLE
[web.archive:youtube] Improve extractor

### PR DESCRIPTION
**Pull request by coletdjnz**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [X] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [X] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [X] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [X] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [X] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)
    - **a couple of the regexps used in metadata extraction are from old youtube-dl commits**

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [X] Improvement
- [ ] New extractor
- [ ] New feature

---

#### IMPROVEMENTS
- Improved metadata extraction 
    - currently best for polymer UI YouTube captures
    - Title extraction should be fairly stable across almost all YouTube versions
    - Extract `filesize` and `extension` from non-standard archive.org headers
        - `unknown_video` extension should no longer occur
- Get archived thumbnails using CDX API
- Smarter webpage capture downloading 
  - query CDX for only 200 OK response captures
  - Tries multiple captures until it gets an infodict with title
  - order: try oldest polymer UI capture, user provided capture date, oldest capture from CDX, oldest capture web archive redirects to, newest capture web archive redirects to
- New extractor arg: `check_all=captures,thumbnails`
  - `thumbnails` queries CDX API many more times to thumbnails archived under other youtube image servers.
  - `captures` adds all other captures dates returned by CDX API to the capture-try list
 - supports `--ignore-no-formats-error` to allow metadata extraction even if video is not archived

#### ISSUES
- For querying captures of a video watch page, CDX API does not provide captures with extra params in the URL
    - e.g. `http://www.youtube.com/watch?v=AkhihxRKcrs&gl=US&hl=en`
    - probably not much of a problem, still have the fallbacks or date in url given
- yt-dlp has no idea how to sort the thumbnails
    - we really only have `filesize` to work with
- De-duplicating thumbnails does not work if they were archived from different ytimg servers.
- filesize not shown when downloading
     - archive.org doesn't provide `content-length` header for the video files

#### TODO
- [X] Fix description extraction for some older versions of YouTube, if possible
- [X] Fix up error handling to ensure everything is non-fatal
- [X] Find and add more tests for edge cases
- [X] Move video file URL request to be first, so we don't have to check for metadata if not needed?

### CODE REVIEW

File: yt_dlp/extractor/archiveorg.py [hunk 1]
```diff
@@ -278,141 +281,377 @@ class YoutubeWebArchiveIE(InfoExtractor):
...
+
+        video_details = traverse_obj(
+            player_response, 'videoDetails', expected_type=dict, get_all=False, default={})
+
+        microformats = traverse_obj(
+            player_response, ('microformat', 'playerMicroformatRenderer'), expected_type=dict, get_all=False, default={})
+
+        video_title = (
+            video_details.get('title')
+            or YoutubeIE._get_text(microformats, 'title')
```
Comment (pukkandan) [inline yt_dlp/extractor/archiveorg.py:490]: not YoutubeBaseIE?

File: yt_dlp/extractor/archiveorg.py [hunk 2]
```diff
@@ -278,141 +281,377 @@ class YoutubeWebArchiveIE(InfoExtractor):
...
+            uploader_id = self._search_regex(
+                r'(?:user|channel)/([^/]+)', uploader_url or '', 'uploader id', default=None)
+
+        upload_date = unified_strdate(
+            dict_get(microformats, ('uploadDate', 'publishDate'))
+            or search_meta(['uploadDate', 'datePublished'])
+            or self._search_regex(
+                [r'(?s)id="eow-date.*?>(.*?)</span>',
+                 r'(?:id="watch-uploader-info".*?>.*?|["\']simpleText["\']\s*:\s*["\'])(?:Published|Uploaded|Streamed live|Started) on (.+?)[<"\']'],  # @7998520
+                webpage, 'upload date', default=None))
```
Comment (pukkandan) [inline yt_dlp/extractor/archiveorg.py:534]: what are those comments?

File: yt_dlp/extractor/archiveorg.py [hunk 3]
```diff
@@ -278,141 +281,377 @@ class YoutubeWebArchiveIE(InfoExtractor):
...
+            for server in (self._YT_ALL_THUMB_SERVERS if try_all else self._YT_DEFAULT_THUMB_SERVERS) for ext in (('jpg', 'webp') if try_all else ('jpg',))]
+
+        thumbnails = []
+        for url in thumbnail_base_urls:
+            response = self._call_cdx_api(
+                video_id, url, filters=['mimetype:image/(?:webp|jpeg)'],
+                collapse=['urlkey'], query={'matchType': 'prefix'})
+            if not response:
+                continue
+            # TODO fix sorting
```
Comment (pukkandan) [inline yt_dlp/extractor/archiveorg.py:561]: If the thumbnail names are like youtube, you can instead copy what I did for youtube (see youtube.py#L2802-2817). Otherwise, nothing we can do but to depend on filesize. u can set `preference` to filesize in that case.

File: yt_dlp/extractor/archiveorg.py [hunk 4]
```diff
@@ -278,141 +281,377 @@ class YoutubeWebArchiveIE(InfoExtractor):
...
+        self.write_debug('Captures to try: ' + ', '.join(str(i) for i in capture_dates if i is not None))
+        info = {'id': video_id}
+        for capture in capture_dates:
+            if not capture:
+                continue
+            webpage = self._download_webpage(
+                (self._WAYBACK_BASE_URL + 'http://www.youtube.com/watch?v=%s') % (capture, video_id),
+                video_id=video_id, fatal=False, errnote='unable to download capture webpage (it may not be archived)',
+                note='Downloading capture webpage')
+            _info_dict = self._extract_metadata(video_id, webpage or '')
```
Comment (pukkandan) [inline yt_dlp/extractor/archiveorg.py:627]: maybe `current_info`?

**Review by pukkandan — APPROVED** [review]
You can merge when u think its ready. Don't sweat too much about the thumbnails. We can fix its sorting separately later if needed

File: yt_dlp/extractor/archiveorg.py (hunk 2 above)
Comment (coletdjnz) [reply yt_dlp/extractor/archiveorg.py:534]: commit they are sourced from. thought it might be useful, idk
